src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs

```diff
@@ -2757,7 +2757,14 @@ public override InterfaceMapping GetInterfaceMap([DynamicallyAccessedMembers(Dyn
                 MethodBase? rtTypeMethodBase = GetMethodBase(reflectedType, classRtMethodHandle);
                 // a class may not implement all the methods of an interface (abstract class) so null is a valid value
                 Debug.Assert(rtTypeMethodBase is null || rtTypeMethodBase is RuntimeMethodInfo);
-                im.TargetMethods[i] = (MethodInfo)rtTypeMethodBase!;
+                RuntimeMethodInfo? targetMethod = rtTypeMethodBase as RuntimeMethodInfo;
+                // the TargetMethod provided to us by runtime internals may be a generic method instance,
+                //  potentially with invalid arguments. TargetMethods in the InterfaceMap should never be
+                //  instances, only definitions.
+                if ((targetMethod?.IsGenericMethod == true) && !targetMethod.IsGenericMethodDefinition)
+                    targetMethod = (RuntimeMethodInfo)targetMethod.GetGenericMethodDefinition();
+
+                im.TargetMethods[i] = targetMethod!;
```

Comment:
```suggestion
                im.TargetMethods[i] = (targetMethod is { IsGenericMethod: true, IsGenericMethodDefinition: false })
                    ? targetMethod.GetGenericMethodDefinition() : targetMethod!;
```
The `(RuntimeMethodInfo)` cast is unnecessary. Also, the condition can be written in more modern and compact style.